### PR TITLE
PS-670-rendition-video_enhance2

### PR DESCRIPTION
**Pull request by jygaulier**
allow to use twig templates in ffmpeg module

### CODE REVIEW

File: databox/api/src/Asset/Attribute/TemplateResolver.php [hunk 1]
```diff
@@ -17,10 +17,13 @@ public function __construct()
         ]);
     }
 
-    public function resolve(string $template, array $values): string
+    public function resolve($template, array $values): string
```
Comment (4rthem) [inline databox/api/src/Asset/Attribute/TemplateResolver.php:20]: Please keep strong type hint here as the type check should be outside.

File: databox/api/src/Asset/Attribute/TemplateResolver.php [hunk 2]
```diff
@@ -17,10 +17,13 @@ public function __construct()
...
     {
-        $template = $this->twig->createTemplate($template);
+        // trivial: template must contain "{{" or "{%" - not escaped by "\" -
+        if (is_string($template) && 1 === preg_match('/(^{|[^\\\\]{)[{%]/', $template)) {
```
Comment (4rthem) [inline databox/api/src/Asset/Attribute/TemplateResolver.php:23]:
```suggestion
        if (str_contains($template, '{')) {
```

File: lib/php/rendition-factory/src/Config/ModuleOptionsResolver.php [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+                    throw $e;
+                } catch (\Exception $e) {
+                    // not enough context to evaluate ? just ignore this option for now
+                }
+            }
+        }
+
+        return $r;
+    }
+}
```
Comment (4rthem) [inline lib/php/rendition-factory/src/Config/ModuleOptionsResolver.php:59]: You are mixing procedural and OOP.
The service should not keep options in its context. You should have only one method:

```php
public function resolveOptions(array $option, array $context): array
```

**Review by 4rthem — CHANGES_REQUESTED** [review]

File: databox/api/src/Asset/Attribute/TemplateResolver.php (hunk 1 above)
Comment (jygaulier) [reply databox/api/src/Asset/Attribute/TemplateResolver.php:20]: the template can be a number

File: databox/api/src/Asset/Attribute/TemplateResolver.php (hunk 2 above)
Comment (jygaulier) [reply databox/api/src/Asset/Attribute/TemplateResolver.php:23]: template can be anything

File: databox/api/src/Asset/Attribute/TemplateResolver.php (hunk 1 above)
Comment (4rthem) [reply databox/api/src/Asset/Attribute/TemplateResolver.php:20]: If so, this resolver should not be called

File: lib/php/rendition-factory/src/Config/ModuleOptionsResolver.php (hunk 3 above)
Comment (4rthem) [reply lib/php/rendition-factory/src/Config/ModuleOptionsResolver.php]: 👍

File: databox/api/src/Asset/Attribute/TemplateResolver.php (hunk 1 above)
Comment (jygaulier) [reply databox/api/src/Asset/Attribute/TemplateResolver.php:20]: only the resolver can decide if the template is twig code, so it's already too late.
if we want to **not** call the resolver for a number/bool..., then we must move the "{" test before the call, thus revealing outside the resolver that it internally uses twig syntax.
Or we must implement some kind of resolver->isTemplate() ?

Comment (4rthem) [reply databox/api/src/Asset/Attribute/TemplateResolver.php:20]: > only the resolver can decide if the template is twig code

Why? We can have a `str_contains($template, '{')` in this service (which only accepts string, remember this service is already used by other services), and conditionally call this resolver only if the input is a string.

So you can check if `$o` is a string in `ModuleOptionsResolver`:

```php
$r[$k] = is_string($o) ? $this->templateResolver->resolve($o, $context) : $o;
```

This is principle of responsibilities, ModuleOptionsResolver can deal with any type of values, but TemplateResolver only resolve template strings.

File: lib/php/rendition-factory/src/Config/ModuleOptionsResolver.php [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+        foreach ($options as $k => $o) {
+            if (is_array($o)) {
+                $r[$k] = $this->compile($o, $context);
+            } else {
+                try {
+                    $r[$k] = $this->templateResolver->resolve($o, $context);
+                } catch (SyntaxError|LoaderError $e) {
+                    throw $e;
+                } catch (\Exception $e) {
+                    // not enough context to evaluate ? just ignore this option for now
```
Comment (4rthem) [inline lib/php/rendition-factory/src/Config/ModuleOptionsResolver.php:32]: Why do you want to hide errors here? We should be noticed.

**Review by 4rthem — CHANGES_REQUESTED** [review]

**Review by 4rthem — CHANGES_REQUESTED** [review]
As discussed, better to resolve option one by one

File: lib/php/rendition-factory/src/Transformer/Video/VideoTransformerBase.php [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+namespace Alchemy\RenditionFactory\Transformer\Video;
+
+use Alchemy\RenditionFactory\Config\ModuleOptionsResolver;
+use Alchemy\RenditionFactory\Context\TransformationContextInterface;
+use Alchemy\RenditionFactory\Transformer\Video\FFMpeg\Format\FormatInterface;
+use FFMpeg;
+use Symfony\Component\DependencyInjection\Attribute\AutowireLocator;
+use Symfony\Component\DependencyInjection\ServiceLocator;
+
+readonly class VideoTransformerBase
```
Comment (4rthem) [inline lib/php/rendition-factory/src/Transformer/Video/VideoTransformerBase.php:12]:
```suggestion
abstract readonly class AbstractVideoTransformer
```

File: lib/php/rendition-factory/src/Transformer/Video/VideoTransformerBase.php [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+    protected FFMpeg\FFMpeg $ffmpeg;
+
+    public function __construct(#[AutowireLocator(FormatInterface::TAG, defaultIndexMethod: 'getFormat')] protected ServiceLocator $formats,
+        protected ModuleOptionsResolver $optionsResolver,
+    ) {
+    }
+
+    protected function prepare(array $options, TransformationContextInterface $context): void
+    {
+        $this->context = $context;
```
Comment (4rthem) [inline lib/php/rendition-factory/src/Transformer/Video/VideoTransformerBase.php:27]: It should not be stored.
(Also you would have a "readonly" error at the next pass)

File: lib/php/rendition-factory/src/Transformer/Video/VideoTransformerBase.php [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+        ];
+
+        $format = $this->optionsResolver->resolveOption($options['format'] ?? null, $resolverContext);
+        if (!$format) {
+            throw new \InvalidArgumentException('Missing format');
+        }
+        if (!$this->formats->has($format)) {
+            throw new \InvalidArgumentException(sprintf('Invalid format %s', $format));
+        }
+        $this->format = $format;
```
Comment (4rthem) [inline lib/php/rendition-factory/src/Transformer/Video/VideoTransformerBase.php:40]: :no_entry:

File: lib/php/rendition-factory/src/Transformer/Video/VideoTransformerBase.php [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+
+        $this->outputFormat = $this->formats->get($format);
```
Comment (4rthem) [inline lib/php/rendition-factory/src/Transformer/Video/VideoTransformerBase.php:42]: :no_entry:

File: lib/php/rendition-factory/src/Transformer/Video/VideoTransformerBase.php [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+
+        $extension = $this->optionsResolver->resolveOption($options['extension'] ?? null, $resolverContext);
+        if (null !== $extension) {
+            if (!in_array($extension, $this->outputFormat->getAllowedExtensions())) {
+                throw new \InvalidArgumentException(sprintf('Invalid extension %s for format %s', $extension, $format));
+            }
+        } else {
+            $extension = $this->outputFormat->getAllowedExtensions()[0];
+        }
+        $this->extension = $extension;
```
Comment (4rthem) [inline lib/php/rendition-factory/src/Transformer/Video/VideoTransformerBase.php:52]: :no_entry:

File: lib/php/rendition-factory/src/Transformer/Video/VideoTransformerBase.php [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+
+        $ffmpegOptions = [
+            'timeout' => $this->optionsResolver->resolveOption($options['timeout'] ?? null, $resolverContext),
+            'threads' => $this->optionsResolver->resolveOption($options['threads'] ?? null, $resolverContext),
+            'logger' => $context->getLogger(),
+        ];
+        $this->ffmpeg = FFMpegHelper::createFFMpeg($ffmpegOptions);
```
Comment (4rthem) [inline lib/php/rendition-factory/src/Transformer/Video/VideoTransformerBase.php:59]: Should not be stored in this class (service) but returned in this function to the caller.

File: lib/php/rendition-factory/src/Transformer/Video/VideoTransformerBase.php [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+    }
+
+    protected function log(string $message): void
+    {
+        $this->context->log($message);
+    }
```
Comment (4rthem) [inline lib/php/rendition-factory/src/Transformer/Video/VideoTransformerBase.php:65]: `$context->log()` should be call directly in Modules

**Review by 4rthem — CHANGES_REQUESTED** [review]

File: lib/php/rendition-factory/src/Transformer/Video/ModuleCommonArgsDTO.php [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+
+namespace Alchemy\RenditionFactory\Transformer\Video;
+
+use Alchemy\RenditionFactory\Config\ModuleOptionsResolver;
+use Alchemy\RenditionFactory\Context\TransformationContextInterface;
+use Alchemy\RenditionFactory\Transformer\Video\FFMpeg\Format\FormatInterface;
+use FFMpeg;
+use Symfony\Component\DependencyInjection\ServiceLocator;
+
+class ModuleCommonArgsDTO
```
Comment (4rthem) [inline lib/php/rendition-factory/src/Transformer/Video/ModuleCommonArgsDTO.php:11]: DTO is a readonly object only transferring information, not business logic.
You can either remove the `DTO` from the name of this class or move the logic from the constructor to the `AbstractVideoTransformerBase` class (better solution IMO)

File: lib/php/rendition-factory/src/Transformer/Video/AbstractVideoTransformerBase.php [hunk 13]
```diff
@@ -0,0 +1,16 @@
+<?php
+
+namespace Alchemy\RenditionFactory\Transformer\Video;
+
+use Alchemy\RenditionFactory\Config\ModuleOptionsResolver;
+use Alchemy\RenditionFactory\Transformer\Video\FFMpeg\Format\FormatInterface;
+use Symfony\Component\DependencyInjection\Attribute\AutowireLocator;
+use Symfony\Component\DependencyInjection\ServiceLocator;
+
+abstract readonly class AbstractVideoTransformerBase
```
[the remark this answers is not in the document]
Comment (4rthem) [reply lib/php/rendition-factory/src/Transformer/Video/AbstractVideoTransformerBase.php:10]: Still todo

**Review by 4rthem — CHANGES_REQUESTED** [review]

**Review by 4rthem — APPROVED** [review]

**Pull request merged**